hooks/checkout

```diff
@@ -40,6 +40,36 @@ check_set() {
   fi
 }
 
+
+# Works like GNU timeout, but implementing it here lets it run on macOS agents.
+# Mostly copied from here: https://stackoverflow.com/a/24413646
+timeout() {
```

Comment:
There's a big difference between this and the gnu implementation: this one is a function, not a command, so we cannot call it like `timeout foo || exit_code=$?`, as we do in the rest of the script, those places need to be updated to do the `set +e`/`set -e` subshell dance.

Reply:
Why can't we call it like that? Isn't this the same as GNU's `timeout`?

`a.sh`:
```bash
set -euo pipefail

timeout() {
  local -r duration=$1
  shift
  local -r cmd=( "$@" )

  # Run in a subshell to avoid job control messages
  (
    "${cmd[@]}" &
    child=$!
    # Avoid default notification in non-interactive shell for SIGTERM
    trap -- "" SIGTERM
    (
      sleep "$duration"
      kill "$child" 2> /dev/null
    ) &
    wait "$child"
  )

  # Return 124 if a timeout occurred, or the exit code of the command.
  local -r rc=$?
  if [[ "$rc" -eq 143 ]]; then
    return 124
  else
    return "$rc"
  fi
}

timeout 1 /tmp/b.sh || exit_code=$?
echo "$exit_code"
```
`b.sh`:
```bash
#!/usr/bin/env bash
exit 42
```

```
/tmp % bash a.sh 
42
```

Reply:
Because putting a bash function as part of a conditional expression changes the semantics of the code running in it.

See https://docs.canva.tech/style-guides/bash/#use-only-simple-commands-in-conditionals:

The exit-on-failure-of-simple-command option (`set -e`) is globally suppresed inside any conditional context, including the RHS of `!` and LHS of `&&` and `||`.

Note this isn't even the same as `set +e`, it's a special evaluation mode only available in conditionals, so calling `set -e` doesn't reverse it.

Reply:
Ah, I understand now. Thanks for the link. Quirks like this are why I don't get the popularity of `set -euo pipefail`. Instead of failing fast, like it's intended to, it just fails in different and unexpected ways. (Well, **I'm** not expecting them :smile:.) Anyway, I think I'm gonna split this function into a separate script in its own directory and add that to the `$PATH`.